sdk/include/opentelemetry/sdk/resource/resource_detector.h

```diff
@@ -0,0 +1,27 @@
+#pragma once
+
+#include "opentelemetry/nostd/unique_ptr.h"
+#include "opentelemetry/sdk/resource/resource.h"
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace resource
+{
+
+class ResourceDetector
+{
+public:
+  virtual std::unique_ptr<opentelemetry::sdk::resource::Resource> Detect() = 0;
```

Comment:
I think the motivation behind that is to avoid excessive copying of a resource. A resource might be appended to each span. Having a `shared_ptr` (I think a unique pointer will not do) enables us to avoid copying the resource and its attributes.

Reply:
Based on suggestion from @jsuereth ,  `Resources::Merge()` and `Resources::Create()` are modified to return stack allocated `Resource` by value. And this resource is then owned by `sdk::tracer_provider` and passed as reference to span. Please refer to `examples/simple/main.cc` in this PR on how the usage would be, and let me know if any concerns.